test/test_spectral_ops.py

```diff
@@ -43,57 +47,162 @@ def method_fn(t):
+            # torch_fns = (torch.fft.fft, torch.Tensor.fft, scripted_fn, scripted_method_fn)
+            torch_fns = (torch_fn, scripted_fn)
+
+            for iargs in test_args:
+                args = list(iargs)
+                input = args[0]
+                args = args[1:]
+
+                expected = numpy_fn(input.cpu().numpy(), *args)
+                exact_dtype = dtype in (torch.double, torch.complex128)
+                for fn in torch_fns:
+                    actual = fn(input, *args)
+                    self.assertEqual(actual, expected, exact_dtype=exact_dtype)
 
     # Note: NumPy will throw a ValueError for an empty input
     @skipCUDAIfRocm
     @skipCPUIfNoMkl
     @onlyOnCPUAndCUDA
-    @dtypes(torch.complex64, torch.complex128)
+    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
     def test_empty_fft(self, device, dtype):
         t = torch.empty(0, device=device, dtype=dtype)
+        match = r"Invalid number of data points \([-\d]*\) specified"
 
-        if self.device_type == 'cuda':
-            with self.assertRaisesRegex(RuntimeError, "cuFFT error"):
-                torch.fft.fft(t)
-            return
-
-        # CPU (MKL)
-        with self.assertRaisesRegex(RuntimeError, "MKL FFT error"):
+        with self.assertRaisesRegex(RuntimeError, match):
             torch.fft.fft(t)
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.ifft(t)
+
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.irfft(t)
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.hfft(t)
+
+        if not t.is_complex():
+            with self.assertRaisesRegex(RuntimeError, match):
+                torch.fft.rfft(t)
+            with self.assertRaisesRegex(RuntimeError, match):
+                torch.fft.ihfft(t)
+
+    def test_fft_invalid_dtypes(self, device):
+        t = torch.randn(64, device=device, dtype=torch.complex128)
+
+        with self.assertRaisesRegex(RuntimeError, "Expected a real input tensor"):
+            torch.fft.rfft(t)
+
+        with self.assertRaisesRegex(RuntimeError, "Expected a real input tensor"):
+            torch.fft.ihfft(t)
 
-    @dtypes(torch.int64, torch.float32)
-    def test_fft_invalid_dtypes(self, device, dtype):
-        if dtype.is_floating_point:
+    @skipCUDAIfRocm
+    @skipCPUIfNoMkl
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.int8, torch.float, torch.double, torch.complex64, torch.complex128)
+    def test_fft_type_promotion(self, device, dtype):
+        if dtype.is_complex or dtype.is_floating_point:
             t = torch.randn(64, device=device, dtype=dtype)
         else:
             t = torch.randint(-2, 2, (64,), device=device, dtype=dtype)
 
-        with self.assertRaisesRegex(RuntimeError, "Expected a complex tensor"):
-            torch.fft.fft(t)
+        PROMOTION_MAP = {
+            torch.int8: torch.complex64,
```

Comment:
This may need a tweak depending on what we decided for int promotions (per above comment).